### PR TITLE
scala 2.12.0

### PR DESCRIPTION
**Pull request by dwijnand**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

### CODE REVIEW

**Comment by ilovezfs** [comment]
The stable block should be eliminated since there's only a stable spec now, no devel or head.

**Comment by dwijnand** [comment]
It's interesting how `audit --strict` didn't catch that.

Dropping the stable block required changing the order of things.

**Comment by dunn** [comment]
There's a SHA mismatch for the docs:
```
==> Downloading http://www.scala-lang.org/files/archive/scala-docs-2.12.0.zip
Warning: Resource docs reports different SHA256: 2b5cdff12806b246d42c3a36dedc4b38f2ec8f3816266651532354343580e668
Downloaded to: /Users/brew/Library/Caches/Homebrew/scala--docs-2.12.0.zip
SHA256: 30848346e9de48d9ec2338af4d3f5dd4749bfc5c69c4797b618fb591fc906516

File: Formula/scala.rb [hunk 1]
```diff
@@ -1,23 +1,8 @@
...
-      sha256 "73bd44375ebffd5f401950a11d78addc52f8164c30d8528d26c82c1f819cfc16"
-    end
-
-    resource "src" do
-      url "https://github.com/scala/scala/archive/v2.11.8.tar.gz"
-      sha256 "4f11273b4b3c771019253b2c09102245d063a7abeb65c7b1c4519bd57605edcf"
-    end
-  end
+  url "http://www.scala-lang.org/files/archive/scala-2.12.0.tgz"
+  sha256 "9993f6fa9a3c89ee2be41ba66264b853e089686361ceaff44b7997624dad1115"
```
Comment (syhily) [inline Formula/scala.rb]: 85d5bc4763e3ef887c9ea6a73e4f53417f0c6df69df64468890edd32e4702bce

**Comment by syhily** [comment]
@ilovezfs I thought the `scala-2.12.0.tgz` file on remote Jenkins test bot may not correct. We have better remove it and start a new test.

**Comment by ilovezfs** [comment]
The discrepancy needs to be accounted for not papered over by changing the checksum.

**Comment by dwijnand** [comment]
Do you know why it's been failing? I see you tried switching the checksums up yourself, and it failed.

Also, is there a way for homebrew to write out these checksums itself? I seem to remember pacman having something like that.

**Comment by dwijnand** [comment]
@ilovezfs I can consistently sha256 the first package that failed on Jenkins as "d4f485b3f05bbf1224ed5db34d5a5322ce8d1676236757a6cfbf4a11ac59ca22":

```
$ shasum -a 256 scala-2.12.0.tgz
d4f485b3f05bbf1224ed5db34d5a5322ce8d1676236757a6cfbf4a11ac59ca22  scala-2.12.0.tgz

$ curl -Ls https://downloads.lightbend.com/scala/2.12.0/scala-2.12.0.tgz | shasum -a 256
d4f485b3f05bbf1224ed5db34d5a5322ce8d1676236757a6cfbf4a11ac59ca22  -

$ curl -Ls https://downloads.lightbend.com/scala/2.12.0/scala-2.12.0.tgz | shasum -a 256
d4f485b3f05bbf1224ed5db34d5a5322ce8d1676236757a6cfbf4a11ac59ca22  -
```

Do you know why Jenkins and/or homebrew says it should be "392325373cacdd6bc99e90bbd6402403b9c936ec345ee373dfa6faf67259cc92?

**Comment by ilovezfs** [comment]
I changed the checksum because I downloaded it locally and that was the checksum. bump-formula-pr can calculate the main one but that's not doing anything differently from downloading it manually.

**Comment by ilovezfs** [comment]
It probably is a checksum for a garbage file like a 404 page. Notice you're getting the correct checksum locally which is the same one I had added.

**Comment by dwijnand** [comment]
Ok, so (correct me if I'm wrong) the issue blocking this from being merged seems to be on Jenkins, not in the actual change I'm presenting here?

I've downloaded all the artefacts again and they checksum to the same values you committed.

**Comment by ilovezfs** [comment]
Right. We need a URL that works on Jenkins and the correct checksum and then it has to pass the testing.

**Comment by dwijnand** [comment]
From @syhily's comment above, is it worth looking at if there's any stale state on the Jenkins test box?

**Comment by ilovezfs** [comment]
No. It passes fetch --retry

**Comment by ilovezfs** [comment]
@BrewTestBot test this please

**Comment by ilovezfs** [comment]
@dwijnand looks like it decided to cooperate this time. What's strange is that we've now seen 3 different checksums and they all listed the files when I did tar -tf so I think something weird is going on with upstream. The tarball on fossies had a different checksum too: https://fossies.org/linux/misc/scala-2.12.0.tgz/

**Comment by MikeMcQuaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Comment by syhily** [comment]
```bash
brew upgrade
==> Upgrading 2 outdated packages, with result:
scala 2.12.0, vim 8.0.0071
==> Upgrading scala
==> Using the sandbox
==> Downloading https://downloads.lightbend.com/scala/2.12.0/scala-2.12.0.tgz
######################################################################## 100.0%
Error: SHA256 mismatch
Expected: d4f485b3f05bbf1224ed5db34d5a5322ce8d1676236757a6cfbf4a11ac59ca22
Actual: 42be98ff9754518fd5c9a942c94ffba2464667a5e95ed4917e4e95565c96bdfe
Archive: /Users/Yufan/Library/Caches/Homebrew/scala-2.12.0.tgz
To retry an incomplete download, remove the file above.
```

What happened to the scala upstream.

**Comment by ilovezfs** [comment]
LOL. @MikeMcQuaid looks like the checksum is still morphing! Cursed be he who does not heed the do-not-merge label!

**Comment by MikeMcQuaid** [comment]
> do-not-merge label!

Missed that. Reverting.

**Comment by MikeMcQuaid** [comment]
Reverted in https://github.com/Homebrew/homebrew-core/commit/7a76277f98a5798cf5da5c11965276b4e22f6711. Sorry about that @dwijnand.

**Comment by ilovezfs** [comment]
@MikeMcQuaid thanks. Yup, the checksum now morphed to 42be98ff9754518fd5c9a942c94ffba2464667a5e95ed4917e4e95565c96bdfe as @syhily said.

```
Josephs-MacBook-Pro:~ joe$ openssl dgst -sha256 <(/usr/local/opt/curl/bin/curl -s -L https://downloads.lightbend.com/scala/2.12.0/scala-2.12.0.tgz)
SHA256(/dev/fd/63)= 42be98ff9754518fd5c9a942c94ffba2464667a5e95ed4917e4e95565c96bdfe
```

@dwijnand do you mind reporting the volatile checksum situation upstream?

**Comment by dwijnand** [comment]
No, I don't mind. Thanks everyone for the help and sorry for it not being smooth.

**Comment by ilovezfs** [comment]
I dug a bit further, and they have a rather nasty packaging problem. Here are some examples of the differences:

```
Josephs-MacBook-Pro:tmp joe$ diff SA/scala-2.12.0/lib/sc/compiler.properties SB/scala-2.12.0/lib/sc/compiler.properties 
1c1
< #Tue Nov 08 04:07:50 UTC 2016
---
> #Mon Nov 07 03:51:19 UTC 2016
Josephs-MacBook-Pro:tmp joe$ diff SA/scala-2.12.0/lib/sc/interactive.properties SB/scala-2.12.0/lib/sc/interactive.properties 
1c1
< #Tue Nov 08 04:07:50 UTC 2016
---
> #Mon Nov 07 03:51:19 UTC 2016  
Josephs-MacBook-Pro:tmp joe$ diff SA/scala-2.12.0/lib/sc/META-INF/MANIFEST.MF SB/scala-2.12.0/lib/sc/META-INF/MANIFEST.MF 
2c2
< Bnd-LastModified: 1478578078986
---
> Bnd-LastModified: 1478490688777
```

**Comment by dwijnand** [comment]
@ilovezfs @MikeMcQuaid @syhily 

I've found the issue: a nightly job was republishing the 2.12.0 release.

This has now been stopped and archived. The last run was Nov  8 05:25 UTC. See https://github.com/scala/scala-jenkins-infra/issues/198 for more info.

So I think this should be ready now. Please could you re-open this pull request and test it? Thanks.

**Comment by ilovezfs** [comment]
of course

**Comment by dwijnand** [comment]
Ugh.. looks like there might be CDN issues now (my colleague in SF is getting different shas than myself in London..). Specifically, I'm getting (in London, UK):

```
$ curl -s http://downloads.lightbend.com/scala/2.12.0/scala-2.12.0.tgz | shasum -a 256
d4f485b3f05bbf1224ed5db34d5a5322ce8d1676236757a6cfbf4a11ac59ca22  -

$ curl -s https://www.scala-lang.org/files/archive/scala-2.12.0.tgz | shasum -a 256
4eb78f594421bb62e241704c20edd07ecf24954718ea0da3a107f73eb61ef7ca  -
```

where my colleague is declaring (in https://github.com/scala/scala-jenkins-infra/issues/198#issuecomment-259239859) it should be

```
42be98ff9754518fd5c9a942c94ffba2464667a5e95ed4917e4e95565c96bdfe
```

which is what he's getting from downloading them in SF, USA.

If it's ok with you I'll sit on this one for a few days, how it all clears up. Sorry.

**Comment by ilovezfs** [comment]
Yup, that's fine.

**Comment by holograph** [comment]
One down, one to go:
```
$ curl -s http://downloads.lightbend.com/scala/2.12.0/scala-2.12.0.tgz | shasum -a 256
42be98ff9754518fd5c9a942c94ffba2464667a5e95ed4917e4e95565c96bdfe  -
$ curl -s https://www.scala-lang.org/files/archive/scala-2.12.0.tgz | shasum -a 256
4eb78f594421bb62e241704c20edd07ecf24954718ea0da3a107f73eb61ef7ca  -
```

**Comment by leaxoy** [comment]
May be we can use `https://www.scala-lang.org`, not `http://downloads.lightbend.com`

**Comment by ilovezfs** [comment]
@holograph the latter URL you used above isn't valid since there's actually no https support.

```
Josephs-MacBook-Pro:~ joe$ curl -s http://www.scala-lang.org/files/archive/scala-2.12.0.tgz| shasum -a 256
42be98ff9754518fd5c9a942c94ffba2464667a5e95ed4917e4e95565c96bdfe  -
Josephs-MacBook-Pro:~ joe$ curl -s https://www.scala-lang.org/files/archive/scala-2.12.0.tgz| shasum -a 256
4eb78f594421bb62e241704c20edd07ecf24954718ea0da3a107f73eb61ef7ca  -
```

**Comment by ilovezfs** [comment]
@dwijnand https -> http and you should get the same checksum as your colleague.

**Comment by dwijnand** [comment]
Yeah, just checked them all and I think this should be (finally!) good to go.

**Comment by dwijnand** [comment]
Do I need to squash or do you have infrastructure to do that?

**Comment by ilovezfs** [comment]
It would be great if you could, actually.

**Review by ilovezfs — APPROVED** [review]

**Comment by ilovezfs** [comment]
@dwijnand :rocket: Thanks for the PR and your extra effort looking into the upstream packaging issue!

**Pull request closed without merging**